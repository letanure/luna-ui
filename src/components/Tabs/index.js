import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';

const changeToActive = (themeColor) => keyframes`
  from {
    border-bottom-color: transparent;
    opacity: .5;
  }

  to {
    border-bottom-color: ${themeColor};
    opacity: 1;
  }
`;

const TabWrapper = styled.div`
  width: 100%;
  display: flex;
`;

const TabInactive = styled.div`
  padding: 5px 10px;
  font-weight: bold;
  color: #333;
  opacity: .5;
  cursor: pointer;
`;

const TabActive = styled.div`
  padding: 5px 10px;
  color: #333;
  font-weight: bold;
  border-bottom: 1px solid transparent;
  animation: ${(props) => changeToActive(props.theme.tabs[props.kind])} .25s ease forwards;
`;

/**
 * Hint component.
 * E.g.:
 * ```html
 *   <Tabs
 *     value={this.state.value}
 *     onChange={(value) => { this.setState({ value }) }}
 *     tabs={{
 *       selection1: <span>Selection 1</span>,
 *       selection2: <span>Selection 2</span>,
 *       selection3: <span>Selection 3</span>,
 *       selection4: <span>Selection 4</span>,
 *       selection5: <span>Selection 5</span>,
 *       selection6: <span>Selection 6</span>
 *     }}
 *   />
 *   <Tabs.SelectView
 *     value={this.state.value}
 *     views={{
 *       selection1: <TabBodyExample>Content for tab 1</TabBodyExample>,
 *       selection2: <TabBodyExample>Content for tab 2</TabBodyExample>,
 *       selection3: <TabBodyExample>Content for tab 3</TabBodyExample>,
 *       selection4: <TabBodyExample>Content for tab 4</TabBodyExample>,
 *       selection5: <TabBodyExample>Content for tab 5</TabBodyExample>,
 *       selection6: <TabBodyExample>Content for tab 6</TabBodyExample>
 *     }}
 *   />
 * ```
 */
class Tabs extends React.PureComponent {
  constructor(props) {
    super(props);
    this.renderTabs = this.renderTabs.bind(this);
  }

  renderTabs() {
    const { value, tabs, onChange, kind } = this.props;
    return Object.keys(tabs).map((key) => {
      let TabComponent = TabInactive;
      if (key === value) {
        TabComponent = TabActive;
      }
      return <TabComponent
        kind={kind}
        key={key}
        id={key}
        onClick={() => onChange(key)}
      >
        {tabs[key]}
      </TabComponent>;
    });
  }

  render() {
    const { className } = this.props;

    return <TabWrapper className={className}>
      {this.renderTabs()}
    </TabWrapper>;
  }
}

Tabs.SelectView = ({ value, views }) => {
  return views[value];
};

Tabs.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  tabs: PropTypes.object.isRequired,
  kind: PropTypes.string,
  className: PropTypes.string
};

Tabs.defaultProps = {
  kind: 'primary',
  className: ''
};

export default Tabs;
