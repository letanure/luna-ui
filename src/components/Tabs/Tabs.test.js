import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import theme from '../../config/defaultTheme';
import ThemeProvider from '../Theme/index';

import Tabs from './index';

describe('<Tabs />', () => {
  let component;
  let onChange;

  beforeEach(() => {
    onChange = sinon.spy();
    component =
      <ThemeProvider theme={theme}>
        <React.Fragment>
          <Tabs
            value="selection1"
            onChange={onChange}
            tabs={{
              selection1: <span>Selection 1</span>,
              selection2: <span>Selection 2</span>,
              selection3: <span>Selection 3</span>,
              selection4: <span>Selection 4</span>,
              selection5: <span>Selection 5</span>,
              selection6: <span>Selection 6</span>
            }}
          />
          <Tabs.SelectView
            value="selection1"
            views={{
              selection1: <div>Content for tab 1</div>,
              selection2: <div>Content for tab 2</div>,
              selection3: <div>Content for tab 3</div>,
              selection4: <div>Content for tab 4</div>,
              selection5: <div>Content for tab 5</div>,
              selection6: <div>Content for tab 6</div>
            }}
          />
        </React.Fragment>
      </ThemeProvider>
    ;
  });

  test('HandleClick', () => {
    const wrapper = mount(component);
    wrapper.find('#selection3').at(0)
      .simulate('click');
    expect(onChange).toHaveProperty('callCount', 1);
  });

  test('toMatchSnapshot', () => {
    const tree = renderer
      .create(component)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
