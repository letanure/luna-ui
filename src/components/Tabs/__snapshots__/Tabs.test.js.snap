// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> toMatchSnapshot 1`] = `
<div>
  <div
    className="sc-bdVaJa eGJdiE"
  >
    <div
      className="sc-htpNat jEnfzJ"
      id="selection1"
      kind="primary"
      onClick={[Function]}
    >
      <span>
        Selection 1
      </span>
    </div>
    <div
      className="sc-bwzfXH cQrmHx"
      id="selection2"
      kind="primary"
      onClick={[Function]}
    >
      <span>
        Selection 2
      </span>
    </div>
    <div
      className="sc-bwzfXH cQrmHx"
      id="selection3"
      kind="primary"
      onClick={[Function]}
    >
      <span>
        Selection 3
      </span>
    </div>
    <div
      className="sc-bwzfXH cQrmHx"
      id="selection4"
      kind="primary"
      onClick={[Function]}
    >
      <span>
        Selection 4
      </span>
    </div>
    <div
      className="sc-bwzfXH cQrmHx"
      id="selection5"
      kind="primary"
      onClick={[Function]}
    >
      <span>
        Selection 5
      </span>
    </div>
    <div
      className="sc-bwzfXH cQrmHx"
      id="selection6"
      kind="primary"
      onClick={[Function]}
    >
      <span>
        Selection 6
      </span>
    </div>
  </div>
  <div>
    Content for tab 1
  </div>
</div>
`;
